odoo/addons/base/models/ir_http.py

```diff
@@ -170,7 +170,12 @@ def _handle_exception(cls, exception):
                 return serve
 
         # Don't handle exception but use werkzeug debugger if server in --dev mode
-        if 'werkzeug' in tools.config['dev_mode'] and not isinstance(exception, werkzeug.exceptions.NotFound):
+        # Don't intercept Json request to respect the JSON Spec and return exception as json
+        # "The Response is expressed as a single JSON Object, with the following members:
+        #   jsonrpc, result, error, id"
+        if getattr(request, '_request_type', '') != 'json' \
+            and 'werkzeug' in tools.config['dev_mode'] \
+            and not isinstance(exception, werkzeug.exceptions.NotFound):
```

Comment:
Do you see any cases where request/exception handling can pass to `ir.http` without having gone through the initial [`http.Root.dispatch()`](https://github.com/odoo/odoo/blob/12.0/odoo/http.py#L1456)? I don't see any.
So perhaps we can keep the simple form of testing `request._request_type`.

However you should write the test the other way around, so that `_request_type`  will only be tested if the other 2 conditions are True. This way your change can't possibly affect anything outside of `dev_mode`, and you can sleep better ;-)

```suggestion
        if ('werkzeug' in tools.config['dev_mode']
                and not isinstance(exception, werkzeug.exceptions.NotFound)
                and request._request_type != 'json'):
```

nit-picking: I suggest to use brackets around the test to avoid backslashes, and to over-indent the continuation lines to avoid matching the indent of the inner block. Linters usually complain about that, btw.

Reply:
Merci

A priori non, jamais de cas ou _request_type ne serait pas présent, mais vu que c'est une fonction plutot sensible je préférais être certain de pas crasher toutes les v12 pour un cas inconnu ;) Mais si le risque (presque nul) est ok pour toi, c'est ok pour moi ;) Et effectivement en le reculant dans le test de la condition, il y a encore moins de risque ainsi.

J'ai eu un mentor (...toi...) qui m'a tjrs dit d'etre prudent surtout pour fixer un bug qui n'en est pas, et en plus en stable. Peut etre trop dans l'extrème dans ce cas ci. Prudent != overkill

Perso, je preferais le cas d'avant, ou le dev mode aidait le developpeur a debugger, plutot que respecter la spec JSON...
Mais apparement, Anthony n'est pas de cet avis.

Si mon search_read crash, je preferais avoir un pdb à la ligne qui crash pour debugger (avec un Disconnected toast as side effect), qu'une belle traceback en front end.

C'est un peu comme quand tu vas sur un controlleur auth='user' sans être connecté tu recois une exception SessionExpired (vu qu'on handle par l'exception qui te redirige vers l'écran de login). C'est moins joli, mais c'est la réalité du mode dev... et c'est ce que l'on veut a priori... La première fois, tu te poses la question, mais après tu le sais..

Voila j'ai pushé les suggestions, si ok pour toi, elle peut donc etre R+'sser

PS: chez moi le linter n'apprécie pas ton indentation ;) 
"w503 line break before binary operator" est-ce que l'on peut ignorer w503 pour Odoo ?
Vu [cet article](https://www.flake8rules.com/rules/W503.html), je suppose que oui

Reply:
> PS: chez moi le linter n'apprécie pas ton indentation ;)
> "w503 line break before binary operator" est-ce que l'on peut ignorer w503 pour Odoo ?
> Vu [cet article](https://www.flake8rules.com/rules/W503.html), je suppose que oui

Oui si ton flake8 est à jour il y a W504 qui est l'opposé de W503, mais de toute façon les 2 sont ignorés par défaut (et c'est normal je pense car les 2 options peuvent avoir du sens suivant le contexte). Si c'est pas le cas chez toi c'est peut-être que tu utilises `--ignore` au lieu de `--extend-ignore` et donc les ignore par défaut sont oubliés (et donc tu risques d'avoir W503 *et* W504 à un moment)

Par contre c'est [E129](https://www.flake8rules.com/rules/E129.html) pour le niveau d'indentation initial ;-)